engine/servermaster/jobmanager.go

```diff
@@ -177,9 +184,64 @@ func (jm *JobManagerImpl) GetJob(ctx context.Context, req *pb.GetJobRequest) (*p
 		return nil, err
 	}
 
+	// if job status is running, forward the request to jobmaster openapi
+	if masterMeta.StatusCode == frameModel.MasterStatusInit {
+		status, err := getJobDetailFromJobMaster(ctx, masterMeta.Addr, req.Id)
+		return buildPBJobWithStatus(masterMeta, status, err)
+	}
+
 	return buildPBJob(masterMeta)
 }
 
+// TODO: add a timeout here
+func getJobDetailFromJobMaster(ctx context.Context, address string, jobID string) (*pb.Status, error) {
+	queryJobPath := fmt.Sprintf(jobStatusAPIFormat, jobID)
+	// TODO: TLS?
+	cli, err := httputil.NewClient(nil)
+	if err != nil {
+		return nil, derrors.ErrJobManagerNewHTTPClientFail.Wrap(err)
+	}
+	url := fmt.Sprintf("http://%s%s", address, queryJobPath)
+	log.Info("get job detail from job master", zap.String("url", url))
+	resp, err := cli.Get(ctx, url)
+	if err != nil {
+		return nil, derrors.ErrJobManagerGetJobDetailFail.Wrap(err)
+	}
+	log.Debug("job master response", zap.Any("response status", resp.Status), zap.String("url", url))
+	defer resp.Body.Close()
+	// Normal case: 404 means that the jobmaster may exits or hasn't implemented openapi `api/v1/jobs/${jobID}`
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, nil
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, derrors.ErrJobManagerReadRespBodyFail.Wrap(err)
+	}
+
+	log.Debug("job master response", zap.Any("response body", string(body)), zap.String("url", url))
+	// if resp code is not 200, body should have the standard format 'grpc.Status'
+	if resp.StatusCode/100 != 2 {
+		var grpcInnerStatus spb.Status
+		if err := protojson.Unmarshal(body, &grpcInnerStatus); err != nil {
+			return nil, derrors.ErrJobManagerRespFormatUnmatch.Wrap(err)
+		}
+		grpcStatus := status.FromProto(&grpcInnerStatus)
+		details := grpcStatus.Details()
+		if len(details) < 1 {
+			return nil, derrors.ErrJobManagerRespFormatUnmatch.GenWithStackByArgs("details in grpc status is empty")
+		}
+		st, ok := details[0].(*pb.Status)
+		if !ok {
+			return nil, derrors.ErrJobManagerRespFormatUnmatch.GenWithStackByArgs("detail[0] in grpc status is not the enginepb.Status")
+		}
+		return st, nil
```

Comment:
Can this logic be extracted? The whole function seems too long to be readable.

Reply:
Fix it. 
A client object `JobHTTPClient` is used.